Add digest for dex config

TeddyAndrieux opened the pull request:
**Component**:

'salt', 'authentication'
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

#2569 

**Summary**:

- Add simple "salt slots" handling in Kubernetes manifest when using `metalk8s_kubernetes` module
- Add digest of secret to `checksum/config` annotation for Dex pods
- Remove `kubectl rollout restart` command from doc
- Add test for Dex static user password change

---

Fixes: #2569

alexandre-allard committed:
tests: add a get_grain helper and use it everywhere

(cherry picked from commit 1e110a311b7bf11169268611d7f4489b16763f38)

TeddyAndrieux committed:
salt: Add salt module to replace slot in nested dict

Currently Salt 3000.3 only support slots on root of salt states, but in
MetalK8s context we need to render slot in nested dictionnary, for
example in some kubernetes manifests.
This commit add a really simple slot implementation for nested
dictionnary

TeddyAndrieux committed:
salt: Add an helper to get k8s object digest

Simple helper to get a kubernetes object and then compute the digest
from this object, it's mainly needed to use in slot call in manifest to
get the digest from a config file in a config map or a secret

TeddyAndrieux committed:
salt: Add ability to use slots in kubernetes manifest

Use our `format_slots` implementation on the kubernetes manifest so that
we can use simple salt slot in kubernetes manifest nested dictionnary

TeddyAndrieux committed:
charts: re-render the dex charts to use `checksum/config`

We need Dex Pods to restart when secret content change, so use the
`checksum/config` annotation on the Dex Pods. This annotation is equal
to the digest of the secret content

This chart is re-rendered using:

```
./charts/render.py dex --namespace metalk8s-auth charts/dex.yaml \
    --service-config dex metalk8s-dex-config \
    charts/dex/ > salt/metalk8s/addons/dex/deployed/chart.sls
```

Fixes: #2569

TeddyAndrieux committed:
Revert "docs: Add step to restart dex deployment after config changes"

This reverts commit 1fb68044eebd31fe91f08ade69cfa956e3b4e216.

This rollout is not longer needed since it's automatically handled by
the salt states

Refs: #2569

TeddyAndrieux committed:
tests: Refactor Authentication to Dex test

Add single `then` to handle Dex authentication success or fail, so that
we can add a single `then` to check Dex authentication result, and move
this `then` to conftest so that we can use it for other tests (like CSC)

TeddyAndrieux committed:
tests: Handle list element in `{get,set}_dict_element` function

Add ability to give int in the path so that we can take an element of
the list `mydict.0` to get the first element of the list in `mydict`

TeddyAndrieux committed:
tests: Add scenario to test Admin static user password change

Use CSC to change the password for the Admin static user and test login
with this new password on Dex

Refs: #2569

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/bugfix/GH-2569-add-digest-for-dex-config` with contents from `bugfix/GH-2569-add-digest-for-dex-config`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/bugfix/GH-2569-add-digest-for-dex-config origin/development/2.6
 $ git merge origin/bugfix/GH-2569-add-digest-for-dex-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/bugfix/GH-2569-add-digest-for-dex-config
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

gdemonet commented on salt/_modules/metalk8s.py:
```diff
@@ -287,3 +288,40 @@ def check_pillar_keys(keys, refresh=True, pillar=None, raise_error=True):
+            return data
+        if fmt[1] not in slots_callers:
+            log.warning(
+                "Malformed slot, only %s supported: %s",
+                slots_callers.keys(), data
```
The ordering of values makes things weird to read IMHO:

```suggestion
                "Malformed slot '%s': invalid caller, must use one of '%s'",
                data, "', '".join(slots_callers.keys())
```

gdemonet commented on salt/_modules/metalk8s.py:
```diff
@@ -287,3 +288,40 @@ def check_pillar_keys(keys, refresh=True, pillar=None, raise_error=True):
+    if isinstance(data, basestring) and data.startswith('__slot__:'):
+        fmt = data.split(":", 2)
+        if len(fmt) != 3:
+            log.warning("Malformed slot: %s", data)
```
(minor) Could provide the expected format in this log line

TeddyAndrieux replied:
sure

gdemonet commented on salt/_modules/metalk8s_kubernetes.py:
```diff
@@ -351,3 +354,37 @@ def list_objects(kind, apiVersion, namespace='default', all_namespaces=False,
     return [obj.to_dict() for obj in result.items]
+
+
+def get_object_digest(object_key=None, checksum='sha256', *args, **kwargs):
```
Maybe just `key` as an argument?

gdemonet replied:
Oh you're actually using a full colon-separated path on the object (I see `data:config.yaml`, when I was expecting `config.yaml`). Maybe `path` is better then.

TeddyAndrieux replied:
Hmm yes `path` make sense

gdemonet commented on salt/_modules/metalk8s.py:
```diff
@@ -287,3 +288,40 @@ def check_pillar_keys(keys, refresh=True, pillar=None, raise_error=True):
+
+        fun, args, kwargs = salt.utils.args.parse_function(fmt[2])
+
+        return slots_callers[fmt[1]][fun](*args, **kwargs)
```
Shouldn't we catch errors from this expression? Like, what if the function isn't in `__salt__`? What if the function raises ValueError/TypeError because we're calling it wrong? Maybe it's fine to leave `CommandExecutionError`s pass through, but some "classic" errors like the ones mentioned above would benefit from pre-formatted error messages for the user.

TeddyAndrieux replied:
Sure I can, just so you know here the salt implementation of slots https://github.com/saltstack/salt/blob/master/salt/state.py#L2285-L2306

gdemonet commented on salt/_modules/metalk8s_kubernetes.py:
```diff
@@ -351,3 +354,37 @@ def list_objects(kind, apiVersion, namespace='default', all_namespaces=False,
+        raise CommandExecutionError('Unable to find the object')
+
+    if object_key:
+        if not isinstance(object_key, list):
+            object_key = object_key.split(':')
+
+        for key in object_key:
+            try:
+                obj = obj[key]
+            except KeyError:
+                raise CommandExecutionError(
+                    'Unable to find key "{}" in the object'.format(
+                        '.'.join(object_key)
+                    )
+                )
```
Can't we use Salt utils for this?

TeddyAndrieux replied:
Which one ?

gdemonet replied:
[This one](https://github.com/saltstack/salt/blob/master/salt/utils/data.py#L785)?

TeddyAndrieux replied:
Right, I will

gdemonet commented on salt/_modules/metalk8s_kubernetes.py:
```diff
@@ -351,3 +354,37 @@ def list_objects(kind, apiVersion, namespace='default', all_namespaces=False,
+
+def get_object_digest(object_key=None, checksum='sha256', *args, **kwargs):
+    """
+    Helper to get the digest of one kubernetes object or from a specific key
+    of this object (usefull to get the digest of one config from ConfigMap)
```
One single object may not be sufficient in all cases: some Pods will for instance consume configuration data from both a Secret and a ConfigMap. But acceptable to keep as a debt ticket, as we don't need it for Dex.

TeddyAndrieux replied:
The goal of this function is to retrieve one if you need to retrieve several then you should just do several calls on this one (and maybe create another module function to do it if you need to).

To me it's not a debt and not an issue it's just a feature we do not need right now so ....

gdemonet replied:
We're kind of forced to have an extensible method if we need more than one config to hash, as the hash should be stored in a single annotation. OK for the debt topic, let's wait for a need to implement this feature, you're right :)

gdemonet reviewed (commented):
Those slots look cool :100:!

I'm a bit concerned of the flexibility however:
- Hashing string(s) in a slot is a good idea, could be reused for other sources (like local files), but not in the current approach which is tied to K8s
- Getting more than a single K8s object will very likely become a need, as workloads often consume data from both ConfigMap(s) and Secret(s)

TeddyAndrieux commented:
> I'm a bit concerned of the flexibility however:

I don't see why

> * Hashing string(s) in a slot is a good idea, could be reused for other sources (like local files), but not in the current approach which is tied to K8s

Why ? For local file you already have `hashutil.digest_file` from salt you do not need this function to handle it

> * Getting more than a single K8s object will very likely become a need, as workloads often consume data from both ConfigMap(s) and Secret(s)

Then we will just need a new function that do several call on `metalk8s_kubernetes.get_object_digest` but right now it's not needed

bert-e commented:
# History mismatch



Merge commit #ca1c1585a65bec1e0d94c77e78cb9023de1c1002 on the integration branch
`w/2.6/bugfix/GH-2569-add-digest-for-dex-config` is merging a branch which is neither the current
branch `bugfix/GH-2569-add-digest-for-dex-config` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `bugfix/GH-2569-add-digest-for-dex-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/bugfix/GH-2569-add-digest-for-dex-config` with contents from `bugfix/GH-2569-add-digest-for-dex-config`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/bugfix/GH-2569-add-digest-for-dex-config origin/development/2.6
 $ git merge origin/bugfix/GH-2569-add-digest-for-dex-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/bugfix/GH-2569-add-digest-for-dex-config
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

gdemonet reviewed (dismissed):
:shipit:

bert-e commented:
# History mismatch



Merge commit #981403740235cdba7971cc8f7d62d2090b78672c on the integration branch
`w/2.6/bugfix/GH-2569-add-digest-for-dex-config` is merging a branch which is neither the current
branch `bugfix/GH-2569-add-digest-for-dex-config` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `bugfix/GH-2569-add-digest-for-dex-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

alexandre-allard commented on tests/post/features/service_configuration.feature:
```diff
@@ -7,3 +7,14 @@ Feature: Cluster and Services Configurations
         Then we have '3' at 'status.available_replicas' for 'dex' Deployment in namespace 'metalk8s-auth'
+
+  Scenario: Update Admin static user password
+        Given: the Kubernetes API is available
```
```suggestion
        Given the Kubernetes API is available
```

TeddyAndrieux replied:
Ohhh right good catch :+1:

bert-e commented:
# History mismatch



Merge commit #2fa0e1317e6f6057b4f931a273c35e940f02b126 on the integration branch
`w/2.6/bugfix/GH-2569-add-digest-for-dex-config` is merging a branch which is neither the current
branch `bugfix/GH-2569-add-digest-for-dex-config` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `bugfix/GH-2569-add-digest-for-dex-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

TeddyAndrieux commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/119a88df2ad2f3097bed35cc23b96759a52eedbd) did not succeed in branch bugfix/GH-2569-add-digest-for-dex-config.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/f9c24fe7f2d06e2b97a1537122ea8c6520e76996) did not succeed in branch w/2.6/bugfix/GH-2569-add-digest-for-dex-config.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/119a88df2ad2f3097bed35cc23b96759a52eedbd) did not succeed in branch bugfix/GH-2569-add-digest-for-dex-config.




*The following options are set:* **approve**

alexandre-allard commented on tests/post/features/service_configuration.feature:
```diff
@@ -7,3 +7,14 @@ Feature: Cluster and Services Configurations
+        # NOTE: Consider that admin user is the first of the userlist
+        # Update password to "new-password"
+        When we update the CSC 'spec.localuserstore.userlist.0.hash' to '$2y$14$pDe0vj917rR3XJQ5iEZvhuSGoWkZg2/qBN/mMLqwFSz9S7EYcbIpO'
+        Then the control-plane Ingress path '/oidc' is available
+        And pods with label 'app.kubernetes.io/name=dex' are 'Ready'
+        And we are not able to login to Dex as 'admin@metalk8s.invalid' using password 'password'
+        And we are able to login to Dex as 'admin@metalk8s.invalid' using password 'new-password'
```
The corresponding methods only exist in `tests/post/steps/test_authentication.py`, I don't think it'll work.
You should either put them in `tests/conftest.py` to make these `Then` available to all your scenario, or declare them in `tests/post/steps/test_service_configuration.py`.
Tests have passed probably because of the typo you had on the first line `Given: the Kubernetes API is available` (https://github.com/scality/metalk8s/pull/2573#discussion_r430894649)

TeddyAndrieux replied:
Right, cool pytest-bdd feature (typo = success) .... 
Anyway ok should be fixed but I don't know if my test work as he actually never run for the moment

bert-e commented:
# History mismatch



Merge commit #12028b4d4a80582f85874dd74a650204b31be2dd on the integration branch
`w/2.6/bugfix/GH-2569-add-digest-for-dex-config` is merging a branch which is neither the current
branch `bugfix/GH-2569-add-digest-for-dex-config` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `bugfix/GH-2569-add-digest-for-dex-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

TeddyAndrieux committed:
salt: Add dex userlist by default to CSC ConfigMap

To change default user password we need it to be present in the default
CSC ConfigMap so that you not need to get the user ID and other
information from Dex Secret to change it in the CSC ConfigMap

bert-e commented:
# History mismatch



Merge commit #945c524d165c7ec58926729df830cecd3bf71b8f on the integration branch
`w/2.6/bugfix/GH-2569-add-digest-for-dex-config` is merging a branch which is neither the current
branch `bugfix/GH-2569-add-digest-for-dex-config` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `bugfix/GH-2569-add-digest-for-dex-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue GH-2569.

Goodbye teddyandrieux.
